### ISSUE TITLE
Discuss the idea, since we're using JDK 11, to use 'var' for all local fields

### ISSUE THREAD

**Issue by Soupertonic** [gomint/gomint#522]
Java 10 introduced the `var` keyword for _lvalues_, which performs type inference based on the _rvalue_. It is definitely easier to read, less verbose, safes time and typing effort, etc.

### Minor regulations
- Use `var` for local fields whenever the _rvalue_ is clear to the reader or writer
- Do not use `var` for local fields whenever the _rvalue_ is unclear e.g. when calling a function

#### Example
```java
// VALID: We're able to perform type inference ourselves. We know what the type is.
public void foo() {
    var foo = new FooType();
}

// INVALID: We're not able to perform type inference ourselves.
// We don't know what the type is we're working with.
public void foo() {
    var foo = doStuff();
}
```

I'd like to hear your thoughts on this. #rfc

**Comment by enricoangelon** [comment]
Interesting, also looks like javascript and yeah, easier to read but to be honest i don't see any other benefit

**Comment by Soupertonic** [comment]
You won't notice the benefit if you've never actively practiced the `var` thingy. I've been working with Go, Rust and other languages that abandoned the `Type type = new Type();` variable declaration style, and it feels superb.

It doesn't bring you any performance gains or so, it's just for the sake of readability. Hiding the verbosity. Because we humans are parsers too. (*￣3￣)╭

**Comment by enricoangelon** [comment]
For me there is no difference, we can use that format without any problem if there is no performance leaks

**Comment by Creeperface01** [comment]
I can say the same about kotlin.

**Comment by Soupertonic** [comment]
There are no performance leaks, gains, decreases. It's all on compiler and syntax level.

**Comment by Ali1v** [comment]
Is  okay 🤔

**Comment by Soupertonic** [comment]
@geNAZt What are your thoughts on this?

**Comment by Janmm14** [comment]
intellij postfix completion makes typing this also much easier

**Comment by Soupertonic** [comment]
> intellij postfix completion makes typing this also much easier

You mean by doing `new FooType()` and then `.var` which will automatically generate the variable for you?

**Comment by dktapps** [comment]
my 2 cents: this would needlessly pollute the git history and blame if it were applied to existing code. For new code, seems like an OK idea.

**Comment by Soupertonic** [comment]
But that would mean inconsistency. Heavy inconsistency, kinda. And for such refactoring, it be probably just one commit.

**Comment by dktapps** [comment]
@chaplchapl have you ever used git blame? It's a very useful tool for discovering who did a change and why (Chesterton's fence). This kind of change puts a big barrier in your commit history that makes such tools useless, like [this one](https://github.com/pmmp/PocketMine-MP/commit/dd17652acab33cc83edb0c8a513d3d20819266bf), or [this one](https://github.com/pmmp/PocketMine-MP/commit/5499ac620c61ea335c2151a2750b7f6cc6dc7e75).

One of the golden rules of SW engineering that I learned from experience: don't change style just because it looks nice. The benefit of reduced OCD triggers is outweighed by the pain it causes in branch merges, history pollution, and so on.

The only time a change is warranted is when there is _significant_ benefit to a change (enough to outweigh the pain that comes with it), or when the style is already inconsistent to begin with.

**Comment by Janmm14** [comment]
> 
> 
> > intellij postfix completion makes typing this also much easier
> 
> You mean by doing `new FooType()` and then `.var` which will automatically generate the variable for you?

yes, if we agree in the end to not change the code style, this can still help

**Comment by Soupertonic** [comment]
> @chaplchapl have you ever used git blame?

No, I did not. But that coincides with the fact that I rarely needed to look something up who did what and when. Except for one case, when somebody changed something simple to something complex, and I had to blame them. HA! Get it? Git **Blame**?? lol. 😂

Well, if you put it like this, I can see why you would not want to do a commit or two for refactoring purposes. The problem I see is: Either this style or that style. I like consistency and agreeing on two distinct styles seems not good. It's like working on "modern"/updated code and then the other second on the ye olde legacy code.

I'mna close this for good. Remaining with the current style, to not use `var` for local fields. One benefit is definitely that we're preserving backwards compatibility.

**Comment by geNAZt** [comment]
I would like to use new features of new JVM specs. This is a good fit for developer convenience. I agree it shouldn't be implemented in one refactoring but we should start using new specs in new code. I would like to integrate it over time. Thoughts?

**Comment by Soupertonic** [comment]
I'd say, if you'd like to use newer JVM features and integrate them over time, sure why not? I'm still not pleased thinking about having different local variable declaration styles but I think I can give this a pass, for now at least.

**Comment by geNAZt** [comment]
Well we should define a feature set we use in which scenario. Currently this isn't documented at all. We use lambdas, default method on interfaces etc., we should discuss the usage of streams, for loops, optionals/null etc. etc.

**Comment by Soupertonic** [comment]
Oh, yeah, `Optional`s are special. I enjoy having `Optional`s around but I'm also afraid that too many `Optional`s are destroying the "workflow" of implementing feature X by having to check the nullability. It's definitely safer than "raw" nulls, though.

It's a tough topic. Sponge solely uses `Optional`s and I can't tell if they're happy with it or not.

Regarding streams and loops: Yeah! Sure, why not? I like thinking about an API that could look like this:
```java
Player target = server.players().filter(
            Player.BY_NAME.apply(targetName),
            Player.IS_IN_GAMEMODE.apply(GameMode.CREATIVE)).findOne();
```

With "decorator functions":
```java
Player target = server.players().filter(
            Player.byName(targetName),
            Player.isInGameMode(GameMode.CREATIVE)).findOne();
```

Scoped filter class:
```java
// PlayerFilter would scope all the player related filter functions so that the "Player"
// namespace doesn't get poluted.
Player target = server.players().filter(
            PlayerFilter.byName(targetName),
            PlayerFilter.isInGameMode(GameMode.CREATIVE)).findOne();
```

I will not go into detail about how this API works, though. I'd assume that for instance `Player::BY_NAME` would be a `Function` type that takes in a `String` and then returns a `Predicate` so that the `#filter` function can filter by the player's name. The same goes for the `Player::IS_IN_GAMEMODE` variant.

**Comment by geNAZt** [comment]
Does someone know the impacts of optionals/streams? The problem i had in the past is that the performance hit was very noticable, optionals can be a huge memory pressue problem i guess, streams  are the same? Since both are normal java objects without any pooling or?

**Comment by Soupertonic** [comment]
I think streams and optionals would mean a lot of primitive type boxing. I don't know whether FastUtil or Trove will help us there.

**Comment by geNAZt** [comment]
so there is no support for "PrimitiveOptionals"? JDK seems to have this: https://docs.oracle.com/javase/8/docs/api/java/util/OptionalLong.html , https://docs.oracle.com/javase/8/docs/api/java/util/OptionalInt.html etc. etc. so boxing shouldn't be a bigger problem. I just wonder if there are people around measuring the impact of the GC/obj alloc aroun Optional vs null
